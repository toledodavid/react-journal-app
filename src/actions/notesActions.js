import Swal from 'sweetalert2';
import { types } from '../types/types';
import { db } from '../firebase/firebase-config';
import { loadNotes } from '../helpers/loadNotes';
import { fileUpload } from '../helpers/fileUpload';


export const startNewEntryNoteAction = () => {
  return async (dispatch, getState) => {
    const {uid} = getState().auth;
    
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }

    const documentReference = await db.collection(`${uid}/journal/notes`).add(newNote);
    //console.log(documentReference);

    dispatch(activeNoteAction(documentReference.id, newNote));
    dispatch(addNewNoteAction(documentReference.id, newNote));
  }
}

export const activeNoteAction = (id, note) => ({
  type: types.noteActive,
  payload: {
    id,
    ...note
  }
});

export const addNewNoteAction = (idNote, note) => ({
  type: types.notesAddNewEntry,
  payload: {
    id: idNote,
    ...note
  }
});

export const startLoadingNotesAction = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotesAction(notes));
  }
}

export const setNotesAction = (notes) => ({
  type: types.notesLoad,
  payload: notes
});

export const startSaveNoteAction = (note) => {
  return async (dispatch, getState) => {
    const {uid} = getState().auth;

    if (!note.url) {
      delete note.url;
    }

    const noteToFirestore = {...note};
    delete noteToFirestore.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

    dispatch(refreshNoteAction(note.id, note));
    
    Swal.fire('Saved', note.title, 'success');
  }
}

export const refreshNoteAction = (idNote, note) => ({
  type: types.noteUpdated,
  payload: {
    id: idNote,
    note
  }
});

export const startLoadingImageAction = (file) => {
  return async (dispatch, getState) => {
    const {active:note} = getState().notes;

    Swal.fire({
      title: 'Uploading',
      html: 'Please wait...',
      allowOutsideClick: false,
      didOpen: () => {
        Swal.showLoading();
      }
    });

    const fileUrl = await fileUpload(file);

    note.url = fileUrl;
    dispatch(startSaveNoteAction(note));

    Swal.close();
  }
}

export const startDeletingNoteAction = (idNote) => {
  return async (dispatch, getState) => {
    const {uid} = getState().auth;

    await db.doc(`${uid}/journal/notes/${idNote}`).delete();

    dispatch(deleteNoteAction(idNote));
  }
}

export const deleteNoteAction = (idNote) => ({
  type: types.noteDelete,
  payload: idNote
});

export const notesLogoutAction = () => ({
  type: types.notesLogoutCleaning
});