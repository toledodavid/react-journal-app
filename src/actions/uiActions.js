import { types } from '../types/types';


export const setErrorAction = (msgError) => ({
  type: types.uiSetError,
  payload: msgError
});

export const removeErrorAction = () => ({
  type: types.uiRemoveError
});

export const startLoadingAction = () => ({
  type: types.uiStartLoading
});

export const finishLoadingAction = () => ({
  type: types.uiFinishLoading
});