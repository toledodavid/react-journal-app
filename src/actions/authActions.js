import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { finishLoadingAction, startLoadingAction } from './uiActions';


export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {

    dispatch(startLoadingAction());

    firebase.auth().signInWithEmailAndPassword(email, password).then(({user}) => {
      dispatch(loginAction(user.uid, user.displayName));
      dispatch(finishLoadingAction());
    }).catch(err => {
      console.log(err);
      dispatch(finishLoadingAction());
    });
  }
}

export const startRegisterWithEmailPasswordNameAction = (email, password, name) => {
  return (dispatch) => {

    dispatch(startLoadingAction());

    firebase.auth().createUserWithEmailAndPassword(email, password).then( async({user}) => {
      await user.updateProfile({displayName: name});
      dispatch(loginAction(user.uid, user.displayName));
      dispatch(finishLoadingAction());
    }).catch(err => {
      console.log(err);
      dispatch(finishLoadingAction());
    });
  }
}

export const startGoogleLoginAction = () => {
  return (dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider).then(({user}) => {
      dispatch(loginAction(user.uid, user.displayName));
    });
  }
}

export const loginAction = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
});


export const startLogoutAction = () => {
  return async(dispatch) => {
    await firebase.auth().signOut();
    dispatch(logoutAction());
  }
}

export const logoutAction = () => ({
  type: types.logout
});