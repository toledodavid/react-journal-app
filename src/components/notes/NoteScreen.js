import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNoteAction } from '../../actions/notesActions';
import { useForm } from '../../hooks/useForm';
import NotesAppBar from './NotesAppBar';


const NoteScreen = () => {

  const dispatch = useDispatch();

  const {active:note} = useSelector(state => state.notes);
  
  const [formValues, handleInputChange, reset] = useForm(note);

  const {title, body} = formValues;

  const activeNoteId = useRef(note.id);

  useEffect(() => {
    if (note.id !== activeNoteId.current) {
      reset(note);
      activeNoteId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNoteAction(formValues.id, {...formValues}));
  }, [formValues, dispatch]);


  return(
    <div className="notes__main-content">

      <NotesAppBar />

      <div className="notes__content">
        <input
          type="text"
          placeholder="Some awesome title"
          className="notes__title-input"
          name="title"
          value={title}
          onChange={handleInputChange}
          autoComplete="off"
        />

        <textarea
          placeholder="What happened today?"
          name="body"
          value={body}
          onChange={handleInputChange}
          className="notes__textarea">
        </textarea>

        {
          note.url &&
            (
              <div className="notes__image">
                <img src={note.url} alt="imagen" />
              </div>
            )
        }

      </div>
    </div>
  );
}


export default NoteScreen;