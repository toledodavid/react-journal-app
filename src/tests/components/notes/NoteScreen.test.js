import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { activeNoteAction, startDeletingNoteAction} from '../../../actions/notesActions';

import NoteScreen from '../../../components/notes/NoteScreen';



jest.mock('../../../actions/notesActions', () => ({
  activeNoteAction: jest.fn(),
  startDeletingNoteAction: jest.fn()
}));


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {
    uid: 'G6ZtyzKCKcXqHloVli6izS0t0Xg1',
    name: 'Juan'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    notes: [],
    active: {
      id: 1234,
      title: 'hola',
      body: 'mundo',
      date: 0
    }
  }
};

let store = mockStore(initState);
store.dispatch = jest.fn();


describe('Tests in <NoteScreen /> component', () => {

  const wrapper = mount(
    <Provider store={store}>
      <NoteScreen />
    </Provider>
  );

  test('It should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('It should call activeNoteAction', () => {
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        name: 'title',
        value: 'Hello'
      }
    });

    expect(activeNoteAction).toHaveBeenCalled();
    expect(activeNoteAction).toHaveBeenLastCalledWith(1234, {
      body: 'mundo',
      title: 'Hello',
      id: 1234,
      date: 0
    });
  });

});