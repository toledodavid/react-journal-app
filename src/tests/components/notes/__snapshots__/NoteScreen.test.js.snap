// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <NoteScreen /> component It should render correctly 1`] = `
<div
  className="notes__main-content animate__animated animate__fadeIn animate__faster"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 de Agosto 2020
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="hola"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today?"
      value="mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
